src/Features/CSharp/Portable/Completion/SuggestionMode/CSharpSuggestionModeCompletionProvider.cs

```diff
@@ -198,5 +202,44 @@ private ITypeSymbol GetDelegateType(TypeInferenceInfo typeInferenceInfo, Compila
 
             return typeSymbol.GetDelegateType(compilation);
         }
+
+        private bool IsPotentialPatternVariableDeclaration(SyntaxToken token)
+        {
+            var patternSyntax = token.GetAncestor<PatternSyntax>();
+            if (patternSyntax == null)
+            {
+                return false;
+            }
+
+            // Statements containing 'not' cannot be valid variable declarations, e.g. 'e is not int $$' and 'e is not (1 and int $$)'
+            if (patternSyntax.GetAncestorsOrThis(a => a is UnaryPatternSyntax unaryPatternSyntax &&
+                                                      unaryPatternSyntax.PatternOperator.IsKind(SyntaxKind.NotKeyword)).Any())
+            {
+                return false;
+            }
+
+            // Patterns containing 'or' cannot contain valid variable declarations, e.g. 'e is 1 or int $$'
+            var possibleBinaryPattern = patternSyntax.Parent;
+            while (possibleBinaryPattern is BinaryPatternSyntax binaryPatternSyntax)
+            {
+                if (binaryPatternSyntax.PatternOperator.IsKind(SyntaxKind.OrKeyword))
+                {
+                    return false;
+                }
+
+                possibleBinaryPattern = binaryPatternSyntax.Parent;
+            }
```

Comment:
can we merge this and the above, and just walk up the patterns seeing if we have either of these types of patterns?